RFC0012.md

```diff
@@ -0,0 +1,72 @@
+
+## Description
+Part of the "Instant On" initiative relies on establishing a partnership between the resource manager (RM) and the networking library that allows the combination to fully setup the messaging environment prior to spawn of an application's processes. Completing this procedure enables applications to communicate without discovery and exchange of network endpoint information.
+
+There are two new APIs required to enable this support:
+
+1. Preconditioning the application for network operations. This typically involves obtaining a security token that must be used by each application process when communicating to another process in the same job. Some network libraries generate this token algorithmically, while others may need to obtain a token from a central server. Precondition values are typically passed to application processes as environmental variables that are recognized by the network library when initialized by the process. Thus, the _PMIx\_server\_setup\_application_ API takes the application's nspace (plus whatever pmix\_info\_t directives the RM chooses to provide), and returns an array of _pmix\_info\_t_ structures. The return is done via a callback function so the API will not block should the library need to obtain the token from a remote server.
+
+  The structures may consist of any combination of key-value pairs, and the RM shall:
+
+  * if the key consists of _PMIX\_SET\_ENVAR_, then the provided string shall be included in the environment of each spawned application process in this nspace
+
+  * if the key consists of _PMIX\_UNSET\_ENVAR_, then the environment of the application shall be searched, and the provided envar removed if present.
+
+  * all other key-values shall be included in the job-level info proved at process start.
+
+  Note: this API is not network-specific. Thus, as other precondition data is identified in the future, internal support can be extended to ensure all precondition data is included without changing this API.
+
+1. Preparing the local network driver to support an application's processes that are being spawned on that node. The "Instant On" initiative requires that a process know how to communicate to each other process at startup. One method of accomplishing this is to "preload" the local network library with the location of all processes in that application, thus allowing the library to compute the required address information for any process. The _PMIx\_server\_setup\_local\_support_ API is called by the RM prior to fork/exec of any local process from the given application. This is defined as a non-blocking call to allow for operations that might not immediately complete. The RM is not allowed to fork/exec any local process from the specified application until the provided callback function has been executed.
```

Comment:
oh, I should point out that markdown will automatically generate the numbers for a list. The raw file just has everything numbered as '1'